audio/mods/openmpttracker.cpp

```diff
@@ -0,0 +1,192 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "audio/mods/openmpttracker.h"
+
+#ifdef USE_OPENMPT
+
+#include <libopenmpt/libopenmpt.h>
+
+#include "common/inttypes.h"
+#include "common/ptr.h"
+#include "common/stream.h"
+#include "common/textconsole.h"
+#include "common/util.h"
+
+#include "audio/audiostream.h"
+#include "audio/decoders/raw.h"
+
+static size_t memoryReaderRead(void *stream, void *dst, size_t bytes) {
+	Common::SeekableReadStream *reader = (Common::SeekableReadStream *)stream;
+
+	if (!reader) {
+		return 0;
+	}
+
+	uint32 receivedBytes = reader->read(dst, bytes);
+
+	if (receivedBytes < bytes) {
+		return 0;
+	}
+
+	return receivedBytes;
+}
+
+static int memoryReaderSeek(void *stream, int64_t offset, int whence) {
+	Common::SeekableReadStream *reader = (Common::SeekableReadStream *)stream;
+
+	if (!reader) {
+		return -1;
+	}
+
+	bool ret = reader->seek(offset, whence);
+
+	if (ret)
+		return 0;
+
+	return -1;
+}
+
+static int64_t memoryReaderTell(void *stream) {
+	Common::SeekableReadStream *reader = (Common::SeekableReadStream *)stream;
+
+	if (reader) {
+		return reader->pos();
+	}
+
+	return -1;
+}
+
+// End memory wrappper
+
+namespace Audio {
+class OpenMPTTrackerMod : public RewindableAudioStream {
+public:
+	OpenMPTTrackerMod(Common::SeekableReadStream *stream, DisposeAfterUse::Flag disposeAfterUse);
+	~OpenMPTTrackerMod();
+
+
+	// ImpulseTrackerMod functions
+	bool isLoaded() {
+		return _mpt_load_successful;
+	}
+
+	int readBuffer(int16 *buffer, const int numSamples) override {
+		openmpt_module_read_mono(_mod, getRate(), numSamples, buffer);
+
+		return numSamples;
+	}
+
+	bool isStereo() const override {
+		return openmpt_module_get_num_channels(_mod) == 2;
```

Comment:
This returns the number of pattern channels (usually 4 or more), not the number of of output channels. In particualar, the documentation says this:

> The number of pattern channels is completely independent of the number of output channels. libopenmpt can render modules in mono, stereo or quad surround, but the choice of which of the three modes to use must not be made based on the return value of this function, which may be any positive integer amount. Only use this function for informational purposes.